pyqtorch/apply.py

```diff
@@ -57,3 +60,35 @@ def apply_operator(
         map(lambda e: "".join(list(e)), [operator_dims, in_state_dims, out_state_dims])
     )
     return einsum(f"{operator_dims},{in_state_dims}->{out_state_dims}", operator, state)
+
+
+def apply_ope_ope(ope1: torch.Tensor, ope2: torch.Tensor) -> torch.Tensor:
```

Comment:
to me, it seems we can simply do a batch matmul here.however you need to move the batch dim to the first dimension so torch doesnt complain.
`   
        batch_first_perm = (2, 0, 1)
        undo_perm = tuple(argsort(tensor(batch_first_perm)))

        def _batch_first(m: Tensor) -> Tensor:
            return permute(m, batch_first_perm)  # This returns shape (batch_size, 2, 2)

        def _batch_last(m: Tensor) -> Tensor:
            return permute(
                m, undo_perm
            )  # We need to undo the permute since PyQ expects (2, 2, batch_size).

        return _batch_last(torch.bmm(_batch_first(ope1), _batch_first(ope2) )
        `

Reply:
What do you think of this version ?

```python 
def apply_ope_ope(operator_1: torch.Tensor, operator_2: torch.Tensor) -> torch.Tensor:
    """
    Compute the product of two operators.   
    The operators must have the same dimensions.

    Args:
        operator_1 (torch.Tensor): The first operator.
        operator_2 (torch.Tensor): The second operator.

    Returns:
        torch.Tensor: The product of the two operators.

    Raises:
        ValueError: If the number of qubits or the batch size differs between the two operators.
    """

    # Dimension verifications:
    n_qubits_1 = int(log2(operator_1.size(1)))
    n_qubits_2 = int(log2(operator_2.size(1)))
    batch_size_1 = operator_1.size(-1)
    batch_size_2 = operator_2.size(-1)
    if n_qubits_1 != n_qubits_2:
        raise ValueError("The number of qubit is different between the two operators.")
    if batch_size_1 != batch_size_2:
        raise ValueError("The number of batch is different between the two operators.")
    
    # Permute the batch size on first dimension to allow torch.bmm():
    def batch_first(operator: torch.Tensor) -> torch.Tensor:
        """
        Permute the operator's batch dimension on first dimension.

        Args:
        operator (torch.Tensor): Operator in size [2**n_qubits, 2**qubits,batch_size].

        Returns:
        torch.Tensor: Operator in size [batch_size, 2**n_qubits, 2**qubits].
        """
        batch_first_perm = (2, 0, 1)
        return torch.permute(operator, batch_first_perm) 

    # Undo the permute since PyQ expects tensor.Size([2**n_qubits, 2**qubits,batch_size]):
    def batch_last(operator: torch.Tensor) -> torch.Tensor:
        """
        Permute the operator's batch dimension on last dimension.

        Args:
        operator (torch.Tensor): Operator in size [batch_size,2**n_qubits, 2**qubits].

        Returns:
        torch.Tensor: Operator in size [2**n_qubits, 2**qubits,batch_size].
        """
        undo_perm = (1,2,0)
        return torch.permute(operator, undo_perm) 

    return batch_last(torch.bmm(batch_first(operator_1), batch_first(operator_2)))

Reply:
Yes